### PR TITLE
feat(app): allow robot restarts to track boot ID and timeout

### LINKED ISSUE #6585: bug: App sometimes spins indefinitely waiting for an OT-2 to restart, because of a discovery client race condition
# Overview

When you update an OT-2, the Opentrons App shows a loading spinner while the OT-2 restarts.

Sometimes, that spinner never goes away. Even if the OT-2 has apparently restarted successfully. You need to close and reopen the Opentrons App to continue using the OT-2.

# Steps to reproduce

Either:

* Upgrade a robot from v3.14.0 to v3.20.1. (This reliably triggered the problem on at least 10 robots, at one recent installation.)
* Click the **Restart** button.
* Perform a factory reset and continue by restarting the robot.

# Current behavior

1. App asks update-server to restart by `POST /server/restart`, server responds quickly with an ”ok I’ll restart” message
2. App sets discovery client to fast poll and marks the robot as restart-pending
3. Eventually, the robot-server goes down due to restart
4. By luck, a discovery client poll happens while the robot-server is down, and the app marks the robot as restarting
5. Eventually, the robot-server comes back up
6. A discovery client poll happens, and the app notices that a robot marked as restarting is now healthy, marking it as up

If (4) doesn't happen—for example, if the robot restarts too quickly—then the app will think the robot never began the restart.

# Expected behavior

* The app should be able to eventually detect when an OT-2 has completed a restart, even in the face of an unreliable network or unexpected timing.
* In case the OT-2 never restarts, the spinner should also have a sensible timeout.

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR moves restart tracking logic from a smear across the `robot-admin` and `buildroot` modules into a single epic in the `robot-admin` module. The new epic:

- Keeps existing up/down tracking behavior (watch for robots to go down and come back up)
    - This logic has a known race condition that could result in a restart being "missed"
- Now also tracks [boot ID](https://github.com/Opentrons/opentrons/pull/7404) as an additional "successful restart" avenue
    - Because we're comparing boot IDs, the race condition above does not apply
- Calls the restart "timed out" after 60 seconds, closing the spinner
    - This will resolve and "missed" restarts and drop the user into a state with a warning modal in case there is some actual problem which caused the robot to not come back

Closes #6585 

## Changelog

- feat(app): allow robot restarts to track boot ID and timeout
    - Includes general restarts, like clicking the "Restart" button (full page spinner)
    - Includes restarts initiated during a robot software update (progress animation in update modal)

## Review requests

This is kind of a gnarly test matrix, but here's the behaviors we want to test. The "Restart" button on the robot page is the easiest way to trigger a restart. This works best with a non-Wi-Fi robot so you can pull the cable at any point to simulate failures.

- Successful restart: robot reboots, spinner pops up then goes away
- Unsuccessful restart: robot reboots, spinner pops up, robot never comes back, spinner goes away after timeout
    - Yanking the Ethernet cable after you click "Restart" can simulate this

We also need to test both of these for robot updates:

- Successful update: robot reboots, you end up on a "robot has updated" success modal
- Unsuccessful update: robot reboots, eventually it times out and gives you an error modal
    - To trigger: pull the Ethernet cable after you get to the "your robot is restarting" message

With those behaviors, we have two types of robots to test on:

- v4.2.x robots and higher - expose a `bootId` in their update server's health endpoint
    - Inspect Redux State > `robotAdmin` > `${robotName}` > `restart` > `bootId` to make sure its being tracked
    - Also look at the payloads of `robotAdmin:RESTART_STATUS_CHANGED` actions, there should be a `bootId` in the action that changes the status to `restart-pending`
- Pre-4.2 robots **do not** have a `bootId`
    - We need to test that the restart spinner fallback logic still works 

Finally:

- The restart timeout of ~60~ **changed to 300** seconds (5 minutes) was an arbitrary choice! Is that too short? Too long?
    - Keep an eye on how long restarts take in your testing, especially during updates

If you need refresher on what the buildroot update epics are supposed to do, [check out the app cookbook](https://coda.io/d/Opentrons-App-Cookbook_dkN7IyJ3B21/Robot-Updates_su8Or#_luIeH) which has a mostly complete outline + flowcharts of the app-side logic of the update procedure.

## Risk assessment

Medium. ~I have not tested this on a robot yet, _and_ it required some changes to the robot update logic.~ After robot testing, this PR appears to be doing its job well when combined with #7608 

Mitigating factors:

- Code is well covered by unit tests, though it's very possible I'm missing some edge cases
- Changes to the robot update logic should only affect code that happens _after_ the update has been sent over, verified, and committed

### CODE REVIEW

File: app/src/redux/buildroot/epic.js [hunk 1]
```diff
@@ -84,9 +92,11 @@ const UNABLE_TO_CANCEL_UPDATE_SESSION =
   'Unable to cancel in-progress update session'
 const UNABLE_TO_COMMIT_UPDATE = 'Unable to commit update'
 const UNABLE_TO_RESTART_ROBOT = 'Unable to restart robot'
-const ROBOT_RECONNECTED_WITH_VERSION = 'Robot reconnected with version'
+const ROBOT_RECONNECTED_WITH_VERSION = 'It looks like your robot has version'
```
Comment (amitlissack) [inline app/src/redux/buildroot/epic.js]: Nit: "It looks like..." sounds wishy washy. Are we not sure what the version is?

**Review by amitlissack — APPROVED** [review]
Should be looked at by JS person.

File: app/src/redux/buildroot/epic.js (hunk 1 above)
Comment (mcous) [reply app/src/redux/buildroot/epic.js]: This was me being lazy and trying to use a single error message for two different cases:

- Robot definitely came back, we know what the version is, and it's wrong
- Robot did not come back or came back partially and the restart spinner timed out
    - In this case, the version we have may be stale cached data

You're right in that this error message is unhelpful. I'll rework into two separate errors to be more accurate and straightforward

**Comment by SyntaxColoring** [comment]
> Calls the restart "timed out" after 60 seconds, closing the spinner

This seems pretty tight to me. Anecdotally, I think I've seen restarts take multiple minutes, depending on whether the motor controller board needs to update its firmware, and potentially even how old the SD card is. Admittedly, though, that's measuring from power-on to API server readiness—not update server readiness.

Should we instead set the timeout very very conservatively? Like 10 minutes?

Edit: And the current "Robot is restarting..." dialog copy says "this may take up to 3 minutes," which still seems kind of tight to me.

**Review by sanni-t — APPROVED** [review]

**Pull request merged**